engine/docs/CODE_STRUCTURE.md

```diff
@@ -0,0 +1,127 @@
+# The Defold engine code structure
+
+## Browsing the engine code
+
+To make it easier to find your way around the code, it's good to know about some of our naming conventions.
+
+### Naming conventions
+
+* `res_*.cpp` - Resource types. These are responsible for loading/unloading a certain (compiled) asset type. E.g. [res_sprite.cpp](../gamesys/src/gamesys/resources/res_sprite.cpp)
+
+* `comp_*.cpp` - Component types. Reponsible for creating a component world and instances. Matched with a resource type using the same filename extension. E.g. [comp_sprite.cpp](../gamesys/src/gamesys/components/comp_sprite.cpp)
+
+* `script_*.cpp` - Script modules. These add new Lua modules to the engine. E.g. [script_sprite.cpp](../gamesys/src/gamesys/scripts/script_sprite.cpp)
+  * Exceptions to the rule: `gameobject_script.cpp`, `gui_script.cpp` and `render_script.cpp`
+
+* `*_ddf.proto` - File formats. These are the file formats used for source and compiled content. [sprite_ddf.proto](../gamesys/proto/gamesys/sprite_ddf.proto)
+
+
+### Libraries
+
+This is the list of libraries in the `defold/engine/<library>` folder that are used to make up the engine.
+For the up-to-date list of libraries, see the `ENGINE_LIBS` in [build.py](../../scripts/build.py)
+
+The order of these libraries is important as they determine the possible ways a library may use another library.
+E.g. `render` may use `graphics`, but the reverse order is not allowed.
+
+* testmain - Helper library for unit tests and platform specific functionality.
+* dlib - Our main util library with containers, sockets, I/O, compression etc
+* texc - Our texture compiler. Creates a .jar+.dll for use with bob.jar
+* [modelc](../modelc/README.md) - Our model compiler. Creates a .jar+.dll for use with bob.jar
+* ddf - Our runtime data format. It's a slimmed down version of Protobuf.
+* platform - Handles some platform specific concepts, like a Window.
+* graphics - Graphics layer with e.g. vertex buffers and shader programs. Backends for OpenGL, Vulkan etc.
+* particle - A system for simulating particles. Use by comp_particle.cpp
+* lua - The regular Lua 5.1 implementation.
+* hid - Human Interface Device library that handles gamepads, mouse, keyboards and touch input.
+* input - A higher level input system that maps HID events into input actions
+* physics - A single frontend for our two physics engines Box 2D and Bullet 3D. Includes the source of Box2D
```

Comment:
Users will ask where the source of Bullet 3D resides. We should mention that it is in packages.